### PR TITLE
fix(api): preserve excel percentage formatting in document extractor (#34067)

### PR DESCRIPTION
**Pull request by snesan821**
Fixes #34067

## Summary
Currently, `ExcelExtractor` just grabs the raw float value for percentage cells via `openpyxl`'s `cell.value` (so `90%` becomes `0.9`). This leads to data degradation when users ingest formatted Excel files into their Knowledge Base.

This adds a quick check before the string cast: if the cell is numeric and its `openpyxl` `number_format` contains a `%`, it formats the float back into the correct percentage string. I used `.10g` to handle the formatting so we avoid weird floating-point math artifacts and drop unnecessary trailing zeros.

Also added a mocked unit test for this specific edge case in `test_excel_extractor.py`.

## Screenshots

| Before | After |
|--------|-------|
| Extracted string output: `0.9` | Extracted string output: `90%` |

## Checklist

- [ ] This change requires a documentation update, included: [Dify Document](https://github.com/langgenius/dify-docs)
- [x] I understand that this PR may be closed in case there was no previous discussion or issues. (This doesn't apply to typos!)
- [x] I've added a test for each change that was introduced, and I tried as much as possible to make a single atomic change.
- [ ] I've updated the documentation accordingly.
- [x] I ran `make lint` and `make type-check` (backend) and `cd web && npx lint-staged` (frontend) to appease the lint gods

### CODE REVIEW

File: api/core/rag/extractor/excel_extractor.py [hunk 1]
```diff
@@ -59,7 +59,16 @@ def extract(self) -> list[Document]:
...
+                                    # Check if cell has percentage formatting
+                                    if isinstance(value, (int, float)) and hasattr(cell, "number_format"):
+                                        number_format = cell.number_format or ""
+                                        if "%" in number_format:
+                                            # Convert decimal to percentage (e.g., 0.9 -> 90%)
+                                            value = f"{value * 100:.10g}%"
+                                        else:
+                                            value = str(value)
+                                    else:
+                                        value = str(value)
```
[the remark this answers is not in the document]
Comment (snesan821) [reply api/core/rag/extractor/excel_extractor.py]: Good catch. I've added a regex-based precision parser in the latest commit to handle proper decimal rounding based on the number_format string.

File: api/tests/unit_tests/core/rag/extractor/test_excel_extractor.py [hunk 2]
```diff
@@ -115,3 +116,30 @@ def test_find_header_and_columns_fallback_and_empty_case(self):
...
         assert extractor._find_header_and_columns(empty_sheet) == (0, {}, 0)
+
+    def test_extract_xlsx_preserves_percentage_formatting(self, monkeypatch):
+        """Test that percentage-formatted cells are extracted as percentages, not decimals."""
+        sheet = _FakeSheet(
+            header_rows=[("Item", "Score")],
+            data_rows=[
+                (_FakeCell("Test A"), _FakeCell(0.9, number_format="0%")),
+                (_FakeCell("Test B"), _FakeCell(0.456, number_format="0.0%")),
+                (_FakeCell("Test C"), _FakeCell(123.45, number_format="General")),
```
[the remark this answers is not in the document]
Comment (snesan821) [reply api/tests/unit_tests/core/rag/extractor/test_excel_extractor.py]: Updated. Added specific test cases to verify rounding behavior for both 0% (rounding to 46%) and 0.00% precision.

File: api/core/rag/extractor/excel_extractor.py [hunk 3]
```diff
@@ -88,6 +103,29 @@ def extract(self) -> list[Document]:
...
+        # Match patterns like "0.00%" or "0%" to extract decimal places
+        match = re.search(r"0(\.0+)?%", number_format)
+        if match:
+            decimal_part = match.group(1)
+            if decimal_part:
+                # Count the number of zeros after the decimal point
+                return len(decimal_part) - 1  # Subtract 1 for the decimal point itself
+            else:
+                # No decimal part means 0 precision
+                return 0
```
[the remark this answers is not in the document]
Comment (snesan821) [reply api/core/rag/extractor/excel_extractor.py]: Moved regex compilation to module scope for performance, streamlined the control flow, and updated the precision parser to handle optional # placeholders.

**Comment by snesan821** [comment]
Ready for a final check!

File: api/core/rag/extractor/excel_extractor.py [hunk 4]
```diff
@@ -88,6 +103,29 @@ def extract(self) -> list[Document]:
...
+        Examples:
+            "0%" -> 0
+            "0.0%" -> 1
+            "0.00%" -> 2
+        Returns None if precision cannot be determined.
+        """
+        import re
+
+        # Match patterns like "0.00%" or "0%" to extract decimal places
+        match = re.search(r"0(\.0+)?%", number_format)
```
[the remark this answers is not in the document]
Comment (snesan821) [reply api/core/rag/extractor/excel_extractor.py]: Thanks!

**Comment by snesan821** [comment]
Addressed all automated feedback and static type constraints.

Performance: Extracted the regex compilation to the module level.

Type Safety: Explicitly cast number_format to str() to resolve Pyrefly CI failures on openpyxl descriptor objects.

Edge Cases: Updated formatting logic to correctly support optional decimal placeholders (#) by trimming trailing zeros, mirroring native Excel rendering. Added corresponding unit tests to prevent future regressions.

Ready for a final human review!
